File: tree/257-binary-tree-paths.cpp
// Solution 1: iterative solution using stack
// N = number of nodes
// O(N) in time (each node is visited once)
// O(N) in space (for storing the result array)
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        stack<pair<TreeNode*, string>> stack({make_pair(root, "")});
        vector<string> ans;
        while(!stack.empty()) {
            auto p = stack.top();
            stack.pop();

            TreeNode* node = p.first;
            string & path = p.second;

            if (node == NULL) continue;

            if (node->left == NULL && node->right == NULL) {
                ans.push_back(path + to_string(node->val));
                continue;
            }

            if (node->left != NULL)
                stack.push(make_pair(node->left, path + to_string(node->val) + "->"));
            if (node->right != NULL)
                stack.push(make_pair(node->right, path + to_string(node->val) + "->"));
        }
        return ans;
    }
};

File: practices/leetcode/tree/257-binary-tree-paths.py
# Solution 1: recursive with dfs
# N = number of nodes
# O(N) in time (each node is visited once)
# O(N) in space (for storing the result)
class Solution:
    def binaryTreePaths(self, root):
        def dfs(node, path):
            if not node.left and not node.right:
                ans.append(path + '->' + str(node.val))
                return
            if node.left: dfs(node.left, path + '->' + str(node.val))
            if node.right: dfs(node.right, path + '->' + str(node.val))

        ans = []
        if root is None: return ans
        dfs(root, '')
        return [path[2:] for path in ans]

# Solution 2: recursive, good code
# pattern: post-order
# same complexity as solution 1
class Solution:
    def binaryTreePaths(self, root):
        def findPath(node):
            if node is None:
                return []
            if not node.left and not node.right:
                return [str(node.val)]
            l = findPath(node.left)
            r = findPath(node.right)
            return [str(node.val) + '->' + x for x in l+r]

        return findPath(root)

# Solution 3: iterative
# Please refer to C++ file
